src/core/api/srp_client_api.cpp

```diff
@@ -0,0 +1,171 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the OpenThread SRP client APIs.
+ */
+
+#include "openthread-core-config.h"
+
+#include <openthread/srp_client.h>
+
+#include "common/instance.hpp"
+#include "common/locator-getters.hpp"
+#include "net/srp_client.hpp"
+
+using namespace ot;
+
+#if OPENTHREAD_CONFIG_SRP_CLIENT_ENABLE
+
+otError otSrpClientStart(otInstance *        aInstance,
+                         const otSockAddr *  aServerSockAddr,
```

Comment:
I think the SRP client on a Thread node should automatically discover SRP server(s) from Thread Network Data. 
There should be no need to configure any server address. Beside, a normal Thread device can't provide any interface for such configuring.

This also led us to another question that if there are multiple SRP servers in Thread network data, which server should the SRP client choose, if not all.

Reply:
I considered this but I think its better to keep these as separate/orthogonal modules.

- The idea is that the `Srp::Clinet` class is a pure client implementation providing APIs to set the server address, etc.
- We can have a different module to monitor the Thread network data to discover BRs (providing SRP server functionality) and then start/configure the clinet through its APIs.
- This module/code can be part of next layer (i.e. user of OT) or we can provide it as an internal module in OT (to discover servers, have some selection logic, and then auto-start the client).
- SLAAC management and `Util:Slaac` is a good example of similar model. We give the option to user and provide a OT module to manage SLAAC addresses (based on network data prefixes) automatically (without user involvement) or they can choose not use this and  then manage the SLAAC addresses directly.

Thoughts?

Reply:
I am fine to enhance it with another module to discover BRs providing SRP servers.